File: internal/core/tools/workspace.go
package tools

import (
	"context"
	"encoding/json"
	"fmt"
)

// --- workspace_bash ---

// WorkspaceBashTool executes a bash command in the agent workspace sandbox.
type WorkspaceBashTool struct{}

func (t *WorkspaceBashTool) Definition() ToolDef {
	return ToolDef{
		Name:        "workspace_bash",
		Description: "Execute a bash command in the agent workspace sandbox.",
		Parameters:  json.RawMessage(`{"type":"object","properties":{"command":{"type":"string","description":"The bash command to execute"}},"required":["command"]}`),
	}
}

func (t *WorkspaceBashTool) Execute(ctx context.Context, args json.RawMessage, tctx ToolContext) (json.RawMessage, error) {
	if tctx.Sandbox == nil {
		return nil, fmt.Errorf("no sandbox available")
	}

	var params struct {
		Command string `json:"command"`
	}
	if err := json.Unmarshal(args, &params); err != nil {
		return nil, fmt.Errorf("parsing args: %w", err)
	}

	stdout, stderr, exitCode, err := tctx.Sandbox.Exec(ctx, "bash", "-c", params.Command)
	if err != nil {
		return nil, fmt.Errorf("executing command: %w", err)
	}

	return json.Marshal(map[string]interface{}{
		"stdout":    stdout,
		"stderr":    stderr,
		"exit_code": exitCode,
	})
}

// --- workspace_read_file ---

// WorkspaceReadFileTool reads a file from the agent workspace sandbox.
type WorkspaceReadFileTool struct{}

func (t *WorkspaceReadFileTool) Definition() ToolDef {
	return ToolDef{
		Name:        "workspace_read_file",
		Description: "Read a file from the agent workspace sandbox.",
		Parameters:  json.RawMessage(`{"type":"object","properties":{"path":{"type":"string","description":"Absolute path to the file in the sandbox"}},"required":["path"]}`),
	}
}

func (t *WorkspaceReadFileTool) Execute(ctx context.Context, args json.RawMessage, tctx ToolContext) (json.RawMessage, error) {
	if tctx.Sandbox == nil {
		return nil, fmt.Errorf("no sandbox available")
	}

	var params struct {
		Path string `json:"path"`
	}
	if err := json.Unmarshal(args, &params); err != nil {
		return nil, fmt.Errorf("parsing args: %w", err)
	}

	content, err := tctx.Sandbox.ReadFile(ctx, params.Path)
	if err != nil {
		return nil, fmt.Errorf("reading file: %w", err)
	}

	return json.Marshal(map[string]string{"content": content})
}

// --- workspace_edit_file ---

// WorkspaceWriteFileTool writes content to a file in the agent workspace sandbox.
type WorkspaceWriteFileTool struct{}

func (t *WorkspaceWriteFileTool) Definition() ToolDef {
	return ToolDef{
		Name:        "workspace_edit_file",
		Description: "Write content to a file in the agent workspace sandbox.",
		Parameters:  json.RawMessage(`{"type":"object","properties":{"path":{"type":"string","description":"Absolute path to the file in the sandbox"},"content":{"type":"string","description":"The content to write"}},"required":["path","content"]}`),
	}
}

func (t *WorkspaceWriteFileTool) Execute(ctx context.Context, args json.RawMessage, tctx ToolContext) (json.RawMessage, error) {
	if tctx.Sandbox == nil {
		return nil, fmt.Errorf("no sandbox available")
	}

	var params struct {
		Path    string `json:"path"`
		Content string `json:"content"`
	}
	if err := json.Unmarshal(args, &params); err != nil {
		return nil, fmt.Errorf("parsing args: %w", err)
	}

	if err := tctx.Sandbox.WriteFile(ctx, params.Path, []byte(params.Content)); err != nil {
		return nil, fmt.Errorf("writing file: %w", err)
	}

	return json.Marshal(map[string]string{"status": "ok"})
}

// RegisterWorkspaceTools registers all built-in workspace tools.
func RegisterWorkspaceTools(registry *Registry) error {
	workspaceTools := []Tool{
		&WorkspaceBashTool{},
		&WorkspaceReadFileTool{},
		&WorkspaceWriteFileTool{},
	}
	for _, t := range workspaceTools {
		if err := registry.Register(t); err != nil {
			return err
		}
	}
	return nil
}

File: src/core/tools/workspace.ts
// ── Workspace tools ────────────────────────────────────────────
// Built-in tools for workspace sandbox interaction.

import type {
	Registry,
	Tool,
	ToolDef,
	ToolContext,
	SandboxRunner,
} from "@aspectrr/beluga-sdk";

class WorkspaceBashTool implements Tool {
	definition(): ToolDef {
		return {
			name: "workspace_bash",
			description:
				"Execute a bash command in the workspace sandbox. Returns stdout, stderr, and exit code.",
			parameters: {
				type: "object",
				properties: {
					command: {
						type: "string",
						description: "The bash command to execute",
					},
				},
				required: ["command"],
			},
		};
	}

	async execute(
		args: Record<string, unknown>,
		ctx: ToolContext,
	): Promise<Record<string, unknown>> {
		if (!ctx.sandbox) throw new Error("no sandbox available for this session");
		const result = await ctx.sandbox.exec(String(args.command));
		return {
			stdout: result.stdout,
			stderr: result.stderr,
			exit_code: result.exitCode,
		};
	}
}

class WorkspaceReadFileTool implements Tool {
	definition(): ToolDef {
		return {
			name: "workspace_read_file",
			description: "Read a file from the workspace sandbox.",
			parameters: {
				type: "object",
				properties: {
					path: {
						type: "string",
						description: "Path to the file (relative to workspace root)",
					},
				},
				required: ["path"],
			},
		};
	}

	async execute(
		args: Record<string, unknown>,
		ctx: ToolContext,
	): Promise<Record<string, unknown>> {
		if (!ctx.sandbox) throw new Error("no sandbox available for this session");
		const content = await ctx.sandbox.readFile(String(args.path));
		return { content };
	}
}

class WorkspaceWriteFileTool implements Tool {
	definition(): ToolDef {
		return {
			name: "workspace_edit_file",
			description: "Write or edit a file in the workspace sandbox.",
			parameters: {
				type: "object",
				properties: {
					path: {
						type: "string",
						description: "Path to the file (relative to workspace root)",
					},
					content: {
						type: "string",
						description: "The content to write",
					},
				},
				required: ["path", "content"],
			},
		};
	}

	async execute(
		args: Record<string, unknown>,
		ctx: ToolContext,
	): Promise<Record<string, unknown>> {
		if (!ctx.sandbox) throw new Error("no sandbox available for this session");
		await ctx.sandbox.writeFile(String(args.path), String(args.content));
		return { success: true };
	}
}

export function registerWorkspaceTools(registry: Registry): void {
	registry.register(new WorkspaceBashTool());
	registry.register(new WorkspaceReadFileTool());
	registry.register(new WorkspaceWriteFileTool());
}
